gridstatus/pjm.py

```diff
@@ -1824,3 +1836,172 @@ def _parse_wind_generation_by_area(self, df: pd.DataFrame):
+            verbose=verbose,
+        )
+
+        return self._parse_dam_as_market_results(df)
+
+    def _parse_dam_as_market_results(self, df: pd.DataFrame):
+        df = df.rename(
+            columns={
+                "locale": "Locale",
+                "service": "Service Type",
+                "mcp": "Market Clearing Price",
+                "mcp_capped": "Market Clearing Price Capped",
+                "as_req_mw": "Ancillary Service Required",
+                "total_mw": "Total MW",
+                "as_mw": "Assigned MW",
+                "ss_mw": "Self-Scheduled MW",
+                "ircmwt2": "Interface Reserve Capability MW",
+                "dsr_as_mw": "Demand Response MW Assigned",
+                "nsr_mw": "Non-Synchronized Reserve MW Assigned",
+            },
+        )
+        df = df[
+            [
+                "Interval Start",
+                "Interval End",
+                "Locale",
+                "Service Type",
+                "Market Clearing Price",
+                "Market Clearing Price Capped",
+                "Ancillary Service Required",
+                "Total MW",
+                "Assigned MW",
+                "Self-Scheduled MW",
+                "Interface Reserve Capability MW",
+                "Demand Response MW Assigned",
+                "Non-Synchronized Reserve MW Assigned",
+            ]
+        ]
+
+        return df.sort_values("Interval Start").reset_index(drop=True)
+
+    @support_date_range(frequency=None)
+    def get_real_time_as_market_results(
+        self,
+        date: str | pd.Timestamp,
+        end: Optional[str | pd.Timestamp] = None,
+        verbose: Optional[bool] = False,
+    ):
+        """
+        Retrieves the real-time ancillary service market results from :
+        https://dataminer2.pjm.com/feed/reserve_market_results/definition
+        Data for the previous day is published daily on business days,
+        typically between 11am and 12pm market time.
+
+        Arguments:
+            date (str or pandas.Timestamp): Start datetime for data
+            end: (str or pandas.Timestamp, optional): End datetime for data.
+                Defaults to one day past `date` if not specified.
+            verbose (bool, optional): print verbose output. Defaults to False.
+
+        Returns:
+            pandas.DataFrame: A DataFrame with real-time ancillary service
+            market results.
+        """
+        if date == "latest":
+            date = "today"
+
+        df = self._get_pjm_json(
+            "reserve_market_results",
+            start=date,
+            params={
+                "fields": "datetime_beginning_ept,datetime_beginning_utc,locale,"
+                "service,mcp,mcp_capped,reg_ccp,reg_pcp,as_req_mw,total_mw,as_mw,"
+                "ss_mw,tier1_mw,ircmwt2,dsr_as_mw,nsr_mw,regd_mw"
+            },
+            end=end,
+            filter_timestamp_name="datetime_beginning",
+            interval_duration_min=5,
```

Comment:
Should this be `60 if date < pd.Timestamp('2022-10-01') else 5`?

> Data posted prior to October 1, 2022 will be hourly granularity. Data posted after October 1st will be five-minute granularity as part of PJM Comprehensive Pricing Reform.

Reply:
Maybe, I'm definitely not clear how to handle this. What if that Oct 1, 2022 line falls between the specified start and end dates?

Reply:
Good point. It doesn't seem like we have resolved this issue yet. My preference would be for two methods, one for data before the frequency change and one for data after to avoid mixing frequencies.

Reply:
Updated so an error is raised if start/end dates that include the granularity change date are specified. As a side note, the granularity actually changes on Sep 1, 2022 when you look at the data, and not the published date of Oct 1, 2022.

![image](https://github.com/user-attachments/assets/81982dff-729b-4667-850d-bf436ca23ea3)